sympy/physics/control/lti.py

```diff
@@ -0,0 +1,553 @@
+
+    """
+    def __new__(cls, num, den, var):
+        num, den = _sympify(num), _sympify(den)
+
+        if not isinstance(var, Symbol):
+            raise TypeError("Variable input must be a Symbol.")
+        if den == 0:
+            raise ValueError("TransferFunction can't have a zero denominator.")
+
+        if (((isinstance(num, Expr) and num.has(Symbol)) or num.is_number) and
+            ((isinstance(den, Expr) and den.has(Symbol)) or den.is_number)):
+                obj = super(TransferFunction, cls).__new__(cls, num, den, var)
+                obj._num = num
+                obj._den = den
+                obj._var = var
+                return obj
+        else:
+            raise TypeError("Unsupported type for numerator or denominator of TransferFunction.")
+
+    @property
+    def num(self):
+        """Returns the Numerator of the Transfer Function.
+
+        Examples
+        ========
+
+        >>> from sympy.abc import s, p
+        >>> from sympy.physics.control.lti import TransferFunction
+        >>> G1 = TransferFunction(s**2 + p*s + 3, s - 4, s)
+        >>> G1.num
+        p*s + s**2 + 3
+        >>> G2 = TransferFunction((p + 5)*(p - 3), (p - 3)*(p + 1), p)
+        >>> G2.num
+        (p - 3)*(p + 5)
+
+        """
+        return self._num
+
+    @property
+    def den(self):
+        """Returns the Denominator of the Transfer Function.
+
+        Examples
+        ========
+
+        >>> from sympy.abc import s, p
+        >>> from sympy.physics.control.lti import TransferFunction
+        >>> G1 = TransferFunction(s + 4, p**3 - 2*p + 4, s)
+        >>> G1.den
+        p**3 - 2*p + 4
+        >>> G2 = TransferFunction(3, 4, s)
+        >>> G2.den
+        4
+
+        """
+        return self._den
+
+    @property
+    def var(self):
+        """Returns the variable that is used to anchor the symbols.
+
+        Examples
+        ========
+
+        >>> from sympy.abc import s, p
+        >>> from sympy.physics.control.lti import TransferFunction
+        >>> G1 = TransferFunction(p**2 + 2*p + 4, p - 6, p)
+        >>> G1.var
+        p
+        >>> G2 = TransferFunction(0, s - 5, s)
+        >>> G2.var
+        s
+
+        """
+        return self._var
+
+    @property
+    def bound_symbols(self):
+        """Returns only variables that are dummy variables.
+
+        Examples
+        ========
+
+        >>> from sympy.abc import s, p, l
+        >>> from sympy.physics.control.lti import TransferFunction
+        >>> G1 = TransferFunction(s + l, s - 2*p, s)
+        >>> G1.bound_symbols
+        [s]
+        >>> G2 = TransferFunction(p**2 + 2*p + 4, p - 6, p)
+        >>> G2.bound_symbols
+        [p]
+
+        """
+        return [self.var]
+
+    def _eval_subs(self, old, new):
+        arg_num = self.args[0].subs(old, new)
+        arg_den = self.args[1].subs(old, new)
+        argnew = TransferFunction(arg_num, arg_den, self.var)
+        return self if old in self.bound_symbols else argnew
```

Comment:
This seems to prevent substituting the `var`. Why?

Reply:
I think Oscar discussed it here: https://github.com/sympy/sympy/pull/19390#issuecomment-640556483
That we should not substitute bound symbols

Reply:
I see that Oscar says that subs shouldn't work on some symbols. If that is the way things are done in sympy, that's fine, but I'm still not sure I understand why subs() and xreplace() should work differently in this case. In fact I'm not sure I know the precise difference in subs and xreplace.

Reply:
The main difference is that `xreplace` is a pure structural replacement whereas `subs` is supposed to represent a mathematically valid substitution. That means that subs will give an unevaluated `Subs` where necessary if a pure replacement wouldn't be mathematically valid:
```julia
In [30]: expr = Derivative((z*f(x)), x)                                                                                           

In [31]: rep = {x:z}                                                                                                              

In [32]: expr.subs(rep).doit() == expr.doit().subs(rep)                                                                           
Out[32]: True

In [33]: expr.xreplace(rep).doit() == expr.doit().xreplace(rep)                                                                   
Out[33]: False
```

Reply:
Thanks! Here is what the commands produce if interested:

```
moorepants@agni:~$ isympy
IPython console for SymPy 1.5.1 (Python 3.7.6-64-bit) (ground types: gmpy)

These commands were executed:
>>> from __future__ import division
>>> from sympy import *
>>> x, y, z, t = symbols('x y z t')
>>> k, m, n = symbols('k m n', integer=True)
>>> f, g, h = symbols('f g h', cls=Function)
>>> init_printing()

Documentation can be found at https://docs.sympy.org/1.5.1/


In [1]: QApplication: invalid style override passed, ignoring it.                                                                           
    Available styles: Windows, Fusion
In [1]: expr = Derivative((z*f(x)), x)                                                                                                      

In [2]: rep = {x:z}                                                                                                                         

In [3]: expr.subs(rep)                                                                                                                      
Out[3]: 
⎛∂         ⎞│   
⎜──(z⋅f(x))⎟│   
⎝∂x        ⎠│x=z

In [4]: expr.subs(rep).doit()                                                                                                               
Out[4]: 
  d       
z⋅──(f(z))
  dz      

In [5]: expr.xreplace(rep)                                                                                                                  
Out[5]: 
d         
──(z⋅f(z))
dz        

In [6]: expr.xreplace(rep).doit()                                                                                                           
Out[6]: 
  d              
z⋅──(f(z)) + f(z)
  dz             
```

Reply:
We need to think about what subs and xreplace should do for transfer functions and if that should be different.

Reply:
> We need to think about what subs and xreplace should do for transfer functions and if that should be different.

I think that is pretty simple:
`xreplace` performs structural replacement of one Symbol to another. We can't replace a Symbol to, let's say, a constant. 
``` python
>>> tf = TransferFunction(s**2 + p*s + p, p**3 + 8*p - s, s)
>>> tf.xreplace({s: f})
TransferFunction(f**2 + f*p + p, -f + p**3 + 8*p, f)
>>> _.bound_symbols
[f]
>>> tf.xreplace({s: 5})
*** TypeError: Variable input must be a Symbol.
```
While on the other hand, `subs` can substitute a Symbol in a transfer function to a number or some other symbol.
``` python
>>> tf = TransferFunction(s**2 + p*s + p, p**3 + 8*p - s, s)
>>> tf.subs(p, 3)
TransferFunction(s**2 + 3*s + 3, 51 - s, s)
>>> tf.subs(s, 3)
TransferFunction(p*s + p + s**2, p**3 + 8*p - s, s)
>>> tf.subs(p, r)
TransferFunction(r*s + r + s**2, r**3 + 8*r - s, s)
```

Reply:
>  xreplace performs structural replacement of one Symbol to another. We can't replace a Symbol to, let's say, a constant.

Evaluating a transfer function for specific values of `s` is something that is needed. For example, evaluating a transfer function for `s=0` gives you the steady state gain (zero frequency gain). Taking the limit of the transfer function wrt to `inf` gives the high frequency gain. Evaluating for any purely imaginary number gives you the frequency response.

Reply:
There can be other methods for those things rather than subs. I think the way to think about it with subs is that you are substituting something into an expression but it should still be the same kind of object i.e. if `f` is a matrix than `f.subs(x, y)` should also be a matrix. Here we have a transfer function `tf` and `tf.subs(x, y)` should always be a transfer function.

The normal way to evaluate a function is like `tf(0)` but to me it makes more sense just to provide an `as_expr()` method that gives the expression so that you can do `tf.as_expr().limit(tf.var, oo)` or `tf.as_expr().subs(tf.var, 0)`.

Reply:
I agree, @oscarbenjamin.
In the future (when the "other" functionality will be added), I can either define a `__call__` method or `as_expr()` method.